### PR TITLE
Adding KafkaWriter

### PR DESCRIPTION
**Pull request by utkarshcmu**
KafkaWriter writes into more than one Kafka Topic. It needs the following settings:

1. metadata.broker.list
2. zk.connect
3. serializer.class
4. topics

For example:
"metadata.broker.list" : "10.0.0.1:9092",
"zk.connect" : "10.0.0.2:2181",
"serializer.class" : "kafka.serializer.StringEncoder",
"topics" : "topic1,topic2"
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23discussion_r31865907%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23discussion_r31865923%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23discussion_r31865930%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23discussion_r31865935%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23issuecomment-109550874%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23discussion_r31865952%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23issuecomment-109552991%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23issuecomment-109758520%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23issuecomment-109792310%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23issuecomment-109550874%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Looks%20good%20%21%20A%20few%20minor%20corrections%20would%20be%20nice.%20I%20can%20do%20those%20corrections%20if%20you%20do%20not%20have%20the%20time%20/%20energy%20/%20...%20let%20me%20know%20...%22%2C%20%22created_at%22%3A%20%222015-06-06T08%3A13%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yes%2C%20I%20think%20you%20have%20a%20better%20idea%20about%20the%20project%20structure%20%2C%20so%20it%20might%20be%20far%20quicker%20for%20you%20to%20edit%20these%20things%20as%20compared%20to%20me.%20But%20the%20writer%20works%20and%20writes%20to%20Kafka%20as%20needed.%20I%20tested%20it%20thoroughly.%20Thanks%20for%20the%20quick%20review%20and%20hopefully%20I%20contribute%20more%20in%20the%20future%20on%20this%20project.%20%3A%2B1%3A%20%20%3A%29%20%22%2C%20%22created_at%22%3A%20%222015-06-06T08%3A31%3A46Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1019609%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/utkarshcmu%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20did%20some%20rework%20and%20created%20a%20new%20PR%20%28%23305%29.%20Your%20review%20of%20what%20I%20did%20would%20be%20welcomed%20%21%22%2C%20%22created_at%22%3A%20%222015-06-07T13%3A53%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%2C%20%7B%22body%22%3A%20%22Closed%20as%20this%20is%20replaced%20by%20%23305%22%2C%20%22created_at%22%3A%20%222015-06-07T19%3A41%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%208cb6edaef1d92c28c1c3e74104a2f47c3487a739%20jmxtrans-output/jmxtrans-output-core/pom.xml%206%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23discussion_r31865952%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Version%20should%20be%20defined%20in%20the%20%60dependencyManagement%60%20section%20of%20the%20parent%20pom.%20That%20way%20we%20ensure%20that%20all%20modules%20are%20using%20the%20same%20version%20of%20each%20dependency.%22%2C%20%22created_at%22%3A%20%222015-06-06T08%3A14%3A41Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20jmxtrans-output/jmxtrans-output-core/pom.xml%3AL53-64%22%7D%2C%20%22Pull%208cb6edaef1d92c28c1c3e74104a2f47c3487a739%20jmxtrans-output/jmxtrans-output-core/pom.xml%204%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23discussion_r31865907%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22The%20goal%20of%20the%20%60jmxtrans-output-core%60%20module%20is%20to%20have%20only%20core%20output%20writers%2C%20with%20minimal%20dependencies.%20Could%20you%20move%20the%20output%20writer%20to%20a%20new%20module%2C%20probably%20named%20%60jmxtrans-output-kafka%60%20%3F%22%2C%20%22created_at%22%3A%20%222015-06-06T08%3A07%3A40Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20jmxtrans-output/jmxtrans-output-core/pom.xml%3AL53-64%22%7D%2C%20%22Pull%208cb6edaef1d92c28c1c3e74104a2f47c3487a739%20jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java%20102%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23discussion_r31865935%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20%60isTest%60%20condition%20is%20dubious.%20If%20you%20want%20a%20way%20to%20unit%20test%20this%20class%2C%20you%20should%20provide%20a%20way%20to%20inject%20the%20%60producer%60%20and%20mock%20it.%22%2C%20%22created_at%22%3A%20%222015-06-06T08%3A12%3A29Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java%3AL1-134%22%7D%2C%20%22Pull%208cb6edaef1d92c28c1c3e74104a2f47c3487a739%20jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java%20116%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23discussion_r31865923%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22It%20seems%20that%20all%20those%20getters%20/%20setters%20are%20exposing%20internal%20state.%20You%20could%20probably%20remove%20them.%22%2C%20%22created_at%22%3A%20%222015-06-06T08%3A09%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java%3AL1-134%22%7D%2C%20%22Pull%208cb6edaef1d92c28c1c3e74104a2f47c3487a739%20jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java%2075%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/304%23discussion_r31865930%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Complexity%20of%20this%20method%20is%20quite%20high.%20It%20seems%20we%20can%20at%20least%20extract%20a%20method%20%60writeSingleValue%28%29%60.%22%2C%20%22created_at%22%3A%20%222015-06-06T08%3A11%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java%3AL1-134%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull 8cb6edaef1d92c28c1c3e74104a2f47c3487a739 jmxtrans-output/jmxtrans-output-core/pom.xml 4'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/304#discussion_r31865907'>File: jmxtrans-output/jmxtrans-output-core/pom.xml:L53-64</a></b>
- <a href='https://github.com/gehel'></a> The goal of the `jmxtrans-output-core` module is to have only core output writers, with minimal dependencies. Could you move the output writer to a new module, probably named `jmxtrans-output-kafka` ?
- [ ] <a href='#crh-comment-Pull 8cb6edaef1d92c28c1c3e74104a2f47c3487a739 jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java 116'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/304#discussion_r31865923'>File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java:L1-134</a></b>
- <a href='https://github.com/gehel'></a> It seems that all those getters / setters are exposing internal state. You could probably remove them.
- [ ] <a href='#crh-comment-Pull 8cb6edaef1d92c28c1c3e74104a2f47c3487a739 jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java 75'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/304#discussion_r31865930'>File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java:L1-134</a></b>
- <a href='https://github.com/gehel'></a> Complexity of this method is quite high. It seems we can at least extract a method `writeSingleValue()`.
- [ ] <a href='#crh-comment-Pull 8cb6edaef1d92c28c1c3e74104a2f47c3487a739 jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java 102'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/304#discussion_r31865935'>File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java:L1-134</a></b>
- <a href='https://github.com/gehel'></a> This `isTest` condition is dubious. If you want a way to unit test this class, you should provide a way to inject the `producer` and mock it.
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/304#issuecomment-109550874'>General Comment</a></b>
- <a href='https://github.com/gehel'></a> Looks good ! A few minor corrections would be nice. I can do those corrections if you do not have the time / energy / ... let me know ...
- <a href='https://github.com/utkarshcmu'></a> Yes, I think you have a better idea about the project structure , so it might be far quicker for you to edit these things as compared to me. But the writer works and writes to Kafka as needed. I tested it thoroughly. Thanks for the quick review and hopefully I contribute more in the future on this project. :+1:  :)
- <a href='https://github.com/gehel'></a> I did some rework and created a new PR (#305). Your review of what I did would be welcomed !
- <a href='https://github.com/gehel'></a> Closed as this is replaced by #305
- [ ] <a href='#crh-comment-Pull 8cb6edaef1d92c28c1c3e74104a2f47c3487a739 jmxtrans-output/jmxtrans-output-core/pom.xml 6'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/304#discussion_r31865952'>File: jmxtrans-output/jmxtrans-output-core/pom.xml:L53-64</a></b>
- <a href='https://github.com/gehel'></a> Version should be defined in the `dependencyManagement` section of the parent pom. That way we ensure that all modules are using the same version of each dependency.

<a href='https://www.codereviewhub.com/jmxtrans/jmxtrans/pull/304?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/jmxtrans/jmxtrans/pull/304?approve=1'></a>&nbsp;<a href='https://github.com/jmxtrans/jmxtrans/pull/304'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by gehel** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/jmxtrans/jmxtrans/pull/304'>Refresh</a>.<a href='#crh-update-none'></a>

File: jmxtrans-output/jmxtrans-output-core/pom.xml [hunk 1]
```diff
@@ -53,6 +53,11 @@
 			<artifactId>commons-jexl</artifactId>
 		</dependency>
 		<dependency>
+			<groupId>org.apache.kafka</groupId>
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/pom.xml]: The goal of the `jmxtrans-output-core` module is to have only core output writers, with minimal dependencies. Could you move the output writer to a new module, probably named `jmxtrans-output-kafka` ?

File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			e.printStackTrace();
+			log.error("Unable to write to kafka. Please debug.");
+		}
+	}
+	
+	public Producer<String, String> getProducer() {
+		return producer;
+	}
+
+	public void setProducer(Producer<String, String> producer) {
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java]: It seems that all those getters / setters are exposing internal state. You could probably remove them.

File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		kafkaProperties.setProperty("zk.connect", Settings.getStringSetting(settings, "zk.connect", null));
+		kafkaProperties.setProperty("serializer.class", Settings.getStringSetting(settings, "serializer.class", null));
+		this.producer= new Producer<String,String>(new ProducerConfig(kafkaProperties));
+		this.topics = Settings.getStringSetting(settings, "topics", null).split(","); 
+	}
+	
+	public void validateSetup(Server server, Query query) throws ValidationException {
+	}
+
+	public void internalWrite(Server server, Query query, ImmutableList<Result> results) throws Exception {
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java]: Complexity of this method is quite high. It seems we can at least extract a method `writeSingleValue()`.

File: jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+								log.debug("Topic: "+topic+" ; Kafka Message: {}", obj.toString());
+								messages.add(new KeyedMessage<String,String>(topic,obj.toString()));
+							}
+						} else {
+							log.warn("Unable to submit non-numeric value to Kafka: [{}] from result [{}]", value, result);
+						}
+					}
+				}
+			}
+			if (!this.isTest)
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/src/main/java/com/googlecode/jmxtrans/model/output/KafkaWriter.java]: This `isTest` condition is dubious. If you want a way to unit test this class, you should provide a way to inject the `producer` and mock it.

**Comment by gehel** [comment]
Looks good ! A few minor corrections would be nice. I can do those corrections if you do not have the time / energy / ... let me know ...

File: jmxtrans-output/jmxtrans-output-core/pom.xml [hunk 5]
```diff
@@ -53,6 +53,11 @@
...
+			<artifactId>kafka_2.10</artifactId>
+			<version>0.8.2.1</version>
```
Comment (gehel) [inline jmxtrans-output/jmxtrans-output-core/pom.xml]: Version should be defined in the `dependencyManagement` section of the parent pom. That way we ensure that all modules are using the same version of each dependency.

**Comment by utkarshcmu** [comment]
Yes, I think you have a better idea about the project structure , so it might be far quicker for you to edit these things as compared to me. But the writer works and writes to Kafka as needed. I tested it thoroughly. Thanks for the quick review and hopefully I contribute more in the future on this project. :+1:  :)

**Comment by gehel** [comment]
I did some rework and created a new PR (#305). Your review of what I did would be welcomed !

**Comment by gehel** [comment]
Closed as this is replaced by #305

**Pull request closed without merging**